components/builder-depot/src/backend/s3.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2016-2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Pkg storage backend variant which uses S3 (or an API compatible clone) for
+//! hart storage.
+//!
+//! Has been tested against AWS S3.
+//!
+//! All packages are stored in a single bucket, using the fully qualified
+//! package ident as the key, prefixed with `pkgs`
+//!
+//! # Configuration
+//!
+//! Currently the S3Handler must be configured with both an access key
+//! ID and a secret access key.
+use std::fs::File;
+use std::io::{BufReader, Read};
+use std::path::{PathBuf, Path};
+use std::str::FromStr;
+
+use hab_core::package::{PackageTarget, Identifiable, PackageIdent};
+use iron::typemap::Key;
+
+use rusoto::{Region, credential::StaticProvider, reactor::RequestDispatcher} ;
+use rusoto_s3::{S3, S3Client, GetObjectRequest, PutObjectRequest, ListObjectsRequest};
+use futures::{Future, Stream};
+
+use DepotUtil;
+use config::S3Cfg;
+
+use error::{Result, Error};
+
+pub struct S3Handler {
+    client: S3Client<StaticProvider, RequestDispatcher>,
+    bucket: String,
+}
+
+impl S3Handler {
+    // The S3 Handler struct contains all of the credential
+    // and target information that we should need to perfom
+    // any backend operations
+    pub fn new(config: S3Cfg) -> Self {
+        let region =
+            match config.endpoint {
+                Some(ref value) => {
+                    Region::Custom { name: "minio_s3".to_owned(), endpoint: value.to_string() }
```

Comment:
Instead of hardcoding the region name here, it might be nicer to just use the aws_region config.  We could put some sane defaults for the minio config for local dev - eg, "minio_s3" and "http://localhost:9000"

Reply:
Yeah I'll have to think about this one. This part of the code hasn't been exercised yet. Right now nothing is actually hardcoded, the way it functions is you can either provide an "endpoint" value to the config or you can provide a "region". The "name" in this case is a bit of boilerplate from the rusoto crate, the endpoint is the actual configured value.

Right now we provide a default value to region and not endpoint. It should be trivial to flip those if we want to effectively default to local mode over S3.

Reply:
Cool, yeah that would be great if we flip over to local mode default.

Reply:
Ok this should be corrected now, minio is the default and this abstraction has been cleaned up a bit!